File: ios/Classes/PianoAnalyticsPlugin.swift
import Flutter
import UIKit

import PianoAnalytics

fileprivate let dateType = UInt8(128)
fileprivate let dateFormatter = ISO8601DateFormatter()

fileprivate class Reader: FlutterStandardReader {
    
    override func readValue(ofType type: UInt8) -> Any? {
        switch type {
        case dateType:
            var value: Int64 = 0
            readBytes(&value, length: 8)
            return Date(timeIntervalSince1970: TimeInterval(value / 1000 ))
        default:
            return super.readValue(ofType: type)
        }
    }
}

fileprivate class ReaderWriter: FlutterStandardReaderWriter {
    
    override func reader(with data: Data) -> FlutterStandardReader {
        Reader(data: data)
    }
}

public class PianoAnalyticsPlugin: NSObject, FlutterPlugin {
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "piano_analytics", binaryMessenger: registrar.messenger(),
            codec: FlutterStandardMethodCodec(readerWriter: ReaderWriter())
        )
        let instance = PianoAnalyticsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            var skipResult = false
            switch call.method {
            // Main
            case "init":
                try handleInit(call)
            case "send":
                try handleSend(call)
            // User
            case "getUser":
                skipResult = true
                try handleGetUser(result)
            case "setUser":
                try handleSetUser(call)
            case "deleteUser":
                PianoAnalytics.shared.deleteUser()
            // Visitor
            case "getVisitorId":
                skipResult = true
                handleGetVisitorId(result)
            case "setVisitorId":
                try handleSetVisitorId(call)
            // Privacy
            case "privacyIncludeStorageFeatures":
                try privacyChangeStorageFeatures(call)
            case "privacyExcludeStorageFeatures":
                try privacyChangeStorageFeatures(call, false)
            case "privacyIncludeProperties":
                try privacyChangeProperties(call)
            case "privacyExcludeProperties":
                try privacyChangeProperties(call, false)
            case "privacyIncludeEvents":
                try privacyChangeEvents(call)
            case "privacyExcludeEvents":
                try privacyChangeEvents(call, false)
            default:
                result(FlutterMethodNotImplemented)
            }
            if !skipResult {
                result(nil)
            }
        } catch let error as PluginError {
            switch error {
            case .message(let message): result(FlutterError.from(call, message))
            case .inner(let inner): result(inner)
            }
        } catch {
            result(FlutterError.from(call, error.localizedDescription))
        }
    }

    private func handleInit(_ call: FlutterMethodCall) throws {
        let arguments = try getArguments(call)
        let visitorIdType = VisitorIdType.from(try getArgument(call, arguments, "visitorIDType"))
        
        var configurationBuilder = ConfigurationBuilder()
            .withSite(try getArgument(call, arguments, "site"))
            .withCollectDomain(try getArgument(call, arguments, "collectDomain"))
            .withVisitorIdType(visitorIdType.rawValue)
        
        if let storageLifetimeVisitor = arguments["storageLifetimeVisitor"] as? Int {
            _ = configurationBuilder.withStorageLifetimeVisitor(storageLifetimeVisitor)
        }
        
        if let visitorStorageMode = arguments["visitorStorageMode"] as? String {
            _ = configurationBuilder.withVisitorStorageMode(try PianoAnalyticsPlugin.getVisitorStorageMode(visitorStorageMode))
        }
        
        if let ignoreLimitedAdvertisingTracking = arguments["ignoreLimitedAdvertisingTracking"] as? Bool {
            _ = configurationBuilder.enableIgnoreLimitedAdTracking(ignoreLimitedAdvertisingTracking)
        }
        
        PianoAnalytics.shared.setConfiguration(configurationBuilder.build())
        
        if let visitorId = arguments["visitorId"] as? String, visitorIdType == .Custom {
            PianoAnalytics.shared.setVisitorId(visitorId)
        }
    }

    private func handleSend(_ call: FlutterMethodCall) throws {
        let arguments = try getArguments(call)
        let events: [[String:Any]] = try getArgument(call, arguments, "events")
        
        PianoAnalytics.shared.sendEvents(
            try events.map { event in
                guard let name = event["name"] as? String else {
                    throw PluginError.message("Undefined event name")
                }
                let properties = try (event["data"] as? [String:Any])?.map { property in
                    guard let value = property.value as? [String:Any] else {
                        throw PluginError.message("Undefined property value of event \"\(property.key)\"")
                    }
                    guard let propertyValue = value["value"] else {
                        throw PluginError.message("Undefined property value of event \"\(property.key)\"")
                    }
                    return try Property.from(property.key, propertyValue, forceType: Property.forceType(value["forceType"] as? String))
                }
                return Event(name, properties: Set(properties ?? []))
            }
        )
    }
    
    private func handleGetUser(_ result: @escaping FlutterResult) throws {
        PianoAnalytics.shared.getUser { user in
            guard let user else {
                result(nil)
                return
            }
            
            result([
                "id": user.id,
                "category": user.category
            ])
        }
    }
    
    private func handleSetUser(_ call: FlutterMethodCall) throws {
        let arguments = try getArguments(call)
        
        PianoAnalytics.shared.setUser(
            try getArgument(call, arguments, "id"),
            category: arguments["category"] as? String,
            enableStorage: arguments["enableStorage"] as? Bool ?? true
        )
    }
    
    private func handleGetVisitorId(_ result: @escaping FlutterResult) {
        PianoAnalytics.shared.getVisitorId { visitorId in
            result(visitorId)
        }
    }
    
    private func handleSetVisitorId(_ call: FlutterMethodCall) throws {
        let arguments = try getArguments(call)
        let visitorId: String = try getArgument(call, arguments, "visitorId")
        
        PianoAnalytics.shared.getConfiguration(.VisitorIdType) { type in
            if type == VisitorIdType.Custom.rawValue {
                PianoAnalytics.shared.setVisitorId(visitorId)
            }
        }
    }
    
    private func privacyChangeStorageFeatures(_ call: FlutterMethodCall, _ include: Bool = true) throws {
        let arguments = try getArguments(call)
        
        let features: [String] = try getArgument(call, arguments, "features")
        let keys = try features.map { try PianoAnalyticsPlugin.getStorageKeyByFeature($0) }
        
        let modes: [String] = try getArgument(call, arguments, "modes")
        let privacyModes = try modes.map { try PianoAnalyticsPlugin.getPrivacyMode($0) }
        
        if include {
            PianoAnalytics.shared.privacyIncludeStorageKeys(keys, privacyModes: privacyModes)
        } else {
            PianoAnalytics.shared.privacyExcludeStorageKeys(keys, privacyModes: privacyModes)
        }
    }
    
    private func privacyChangeProperties(_ call: FlutterMethodCall, _ include: Bool = true) throws {
        let arguments = try getArguments(call)
        
        let propertyNames: [String] = try getArgument(call, arguments, "propertyNames")
        
        let modes: [String] = try getArgument(call, arguments, "modes")
        let privacyModes = try modes.map { try PianoAnalyticsPlugin.getPrivacyMode($0) }
        
        let eventNames = arguments["eventNames"] as? [String]
        
        if include {
            PianoAnalytics.shared.privacyIncludeProperties(propertyNames, privacyModes: privacyModes, eventNames: eventNames)
        } else {
            PianoAnalytics.shared.privacyExcludeProperties(propertyNames, privacyModes: privacyModes, eventNames: eventNames)
        }
    }
    
    private func privacyChangeEvents(_ call: FlutterMethodCall, _ include: Bool = true) throws {
        let arguments = try getArguments(call)
        
        let eventNames: [String] = try getArgument(call, arguments, "eventNames")
        
        let modes: [String] = try getArgument(call, arguments, "modes")
        let privacyModes = try modes.map { try PianoAnalyticsPlugin.getPrivacyMode($0) }
        
        if include {
            PianoAnalytics.shared.privacyIncludeEvents(eventNames, privacyModes: privacyModes)
        } else {
            PianoAnalytics.shared.privacyExcludeEvents(eventNames, privacyModes: privacyModes)
        }
    }
    
    private static func getStorageKeyByFeature(_ feature: String) throws -> String {
        switch feature {
        case "VISITOR":
            return PA.Privacy.Storage.VisitorId
        case "CRASH":
            return PA.Privacy.Storage.Crash
        case "LIFECYCLE":
            return PA.Privacy.Storage.Lifecycle
        case "PRIVACY":
            return PA.Privacy.Storage.Privacy
        case "USER":
            return PA.Privacy.Storage.User
        default:
            throw PluginError.message("Invalid feature \"\(feature)\"")
        }
    }
    
    private static func getPrivacyMode(_ name: String) throws -> String {
        switch name {
        case "opt-in":
            return "optin"
        case "opt-out":
            return "optout"
        case "exempt":
            return "exempt"
        case "custom":
            return "custom"
        case "no-consent":
            return "no-consent"
        case "no-storage":
            return "no-storage"
        default:
            throw PluginError.message("Invalid privacy mode \"\(name)\"")
        }
    }
    
    private static func getVisitorStorageMode(_ name: String) throws -> String {
        switch name {
        case "fixed":
            return "fixed"
        case "relative":
            return "relative"
        default:
            throw PluginError.message("Invalid visitor storage mode \"\(name)\"")
        }
    }
}

fileprivate extension VisitorIdType {
    
    static func from(_ name: String) -> Self {
        switch name {
        case "ADID": return VisitorIdType.IDFA
        case "CUSTOM": return VisitorIdType.Custom
        default: return VisitorIdType.UUID
        }
    }
}

fileprivate extension Property {
    
    static func forceType(_ name: String?) throws -> PA.PropertyType? {
        guard let name else {
            return nil
        }
        
        switch name {
        case "b":
            return .bool
        case "n":
            return .int
        case "f":
            return .float
        case "s":
            return .string
        case "d":
            return .date
        case "a:n":
            return .intArray
        case "a:f":
            return .floatArray
        case "a:s":
            return .stringArray
        default:
            throw PluginError.message("Undefined force type \"\(name)\"")
        }
    }
    
    static func from(_ name: String, _ value: Any, forceType: PA.PropertyType? = nil) throws -> Property {
        switch value {
        case let v as Bool:
            return try Property(name, v, forceType: forceType)
        case let v as Int:
            return try Property(name, v, forceType: forceType)
        case let v as Int64:
            return try Property(name, v, forceType: forceType)
        case let v as Double:
            return try Property(name, v, forceType: forceType)
        case let v as String:
            return try Property(name, v, forceType: forceType)
        case let v as Date:
            return try Property(name, v)
        case let v as [Int]:
            return try Property(name, v, forceType: forceType)
        case let v as [Double]:
            return try Property(name, v, forceType: forceType)
        case let v as [String]:
            return try Property(name, v, forceType: forceType)
        default:
            throw PluginError.message("Undefined type for event property \"\(name)\"")
        }
    }
}

File: android/src/main/kotlin/io/piano/flutter/piano_analytics/PianoAnalyticsPlugin.kt
package io.piano.flutter.piano_analytics

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.common.StandardMethodCodec
import io.piano.android.analytics.Configuration
import io.piano.android.analytics.CustomHttpDataProvider
import io.piano.android.analytics.PianoAnalytics
import io.piano.android.analytics.ReportUrlProvider
import io.piano.android.analytics.model.Event
import io.piano.android.analytics.model.PrivacyMode
import io.piano.android.analytics.model.PrivacyStorageFeature
import io.piano.android.analytics.model.Property
import io.piano.android.analytics.model.PropertyName
import io.piano.android.analytics.model.User
import io.piano.android.analytics.model.VisitorIDType
import io.piano.android.analytics.model.VisitorStorageMode
import io.piano.android.consents.PianoConsents
import java.lang.ref.WeakReference
import java.nio.ByteBuffer
import java.util.Date

private class Codec : StandardMessageCodec() {

    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? = when (type) {
        DATE -> Date(buffer.getLong())
        else -> super.readValueOfType(type, buffer)
    }

    companion object {
        const val DATE = 128.toByte()
    }
}

class HttpDataProvider : CustomHttpDataProvider {

    val headers = mutableMapOf<String, String>()
    var parameters = mutableMapOf<String, String>()

    override fun headers() = headers
    override fun parameters() = parameters
}

class PianoAnalyticsPlugin(
    private val httpDataProvider: HttpDataProvider = HttpDataProvider()
) : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var visitorIDType: VisitorIDType
    private lateinit var reportUrlProvider: MutableReportUrlProvider

    private var context: WeakReference<Context?> = WeakReference(null)

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        context = WeakReference(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        context = WeakReference(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        context = WeakReference(binding.activity)
    }

    override fun onDetachedFromActivity() {
        context = WeakReference(null)
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "piano_analytics",
            StandardMethodCodec(Codec())
        )
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            val methodResult: Any? = when (call.method) {
                // Main
                "init" -> handleInit(call)
                "setHeader" -> handleSetHeader(call)
                "setQuery" -> handleSetQuery(call)
                "send" -> handleSend(call)
                // User
                "getUser" -> handleGetUser()
                "setUser" -> handleSetUser(call)
                "deleteUser" -> PianoAnalytics.getInstance().userStorage.currentUser = null
                // Visitor
                "getVisitorId" -> handleGetVisitorId()
                "setVisitorId" -> handleSetVisitorId(call)
                // Privacy
                "privacyIncludeStorageFeatures" -> privacyChangeStorageFeatures(call)
                "privacyExcludeStorageFeatures" -> privacyChangeStorageFeatures(call, false)
                "privacyIncludeProperties" -> privacyChangeProperties(call)
                "privacyExcludeProperties" -> privacyChangeProperties(call, false)
                "privacyIncludeEvents" -> privacyChangeEvents(call)
                "privacyExcludeEvents" -> privacyChangeEvents(call, false)
                else -> error("Unknown method")
            }
            result.success(methodResult.takeUnless { it is Unit })
        } catch (e: Throwable) {
            result.error(call.method, e.message, null)
        }
    }

    private fun handleInit(call: MethodCall) {
        try {
            PianoAnalytics.getInstance()
            reportUrlProvider.update(
                collectDomain = call.arg("collectDomain"),
                site = call.arg("site")
            )
        } catch (_: IllegalStateException) {
            visitorIDType = getVisitorIDType(call.arg("visitorIDType"))
            call.argument<Map<String, String>>("headers")?.let {
                httpDataProvider.headers.putAll(it)
            }
            call.argument<Map<String, String>>("query")?.let {
                httpDataProvider.parameters.putAll(it)
            }
            reportUrlProvider = MutableReportUrlProvider(
                collectDomain = call.arg("collectDomain"),
                site = call.arg("site")
            )
            val pianoAnalytics = PianoAnalytics.init(
                context = context.get() ?: error("Activity not attached"),
                configuration = Configuration.Builder(
                    reportUrlProvider = reportUrlProvider,
                    visitorIDType = visitorIDType,
                    visitorStorageLifetime = call.argument<Int>("storageLifetimeVisitor")
                        ?: Configuration.DEFAULT_VISITOR_STORAGE_LIFETIME,
                    visitorStorageMode = getVisitorStorageMode(call.argument("visitorStorageMode")),
                    ignoreLimitedAdTracking = call.argument<Boolean>("ignoreLimitedAdvertisingTracking")
                        ?: false,
                ).build(),
                pianoConsents = getPianoConsents(),
                customHttpDataProvider = httpDataProvider
            )

            if (visitorIDType == VisitorIDType.CUSTOM) {
                call.argument<String>("visitorId")?.let {
                    pianoAnalytics.customVisitorId = it
                }
            }
        }
    }

    private fun handleSetHeader(call: MethodCall) {
        val value = call.argument<String>("value")
        if (value != null) {
            httpDataProvider.headers[call.arg("key")] = value
        } else {
            httpDataProvider.headers.remove(call.arg("key"))
        }
    }

    private fun handleSetQuery(call: MethodCall) {
        val value = call.argument<String>("value")
        if (value != null) {
            httpDataProvider.parameters[call.arg("key")] = value
        } else {
            httpDataProvider.parameters.remove(call.arg("key"))
        }
    }

    private fun handleSend(call: MethodCall) {
        val events = call.arg<List<Map<String, Any>>>("events").map {
            val name = it["name"] as? String ?: error("Undefined event name")
            Event.Builder(
                name = name,
                properties = (it["data"] as? Map<*, *>)?.map { property ->
                    val propertyName = property.key as? String
                        ?: error("Undefined property name of event \"$name\"")
                    val value = property.value as? Map<*, *>
                        ?: error("Undefined property value of event \"$propertyName\"")
                    val propertyValue = value["value"]
                        ?: error("Undefined property value of event \"$propertyName\"")
                    val forceType = (value["forceType"] as? String)?.let { type ->
                        Property.Type.entries.firstOrNull { t -> t.prefix == type }
                    }
                    getProperty(
                        propertyName,
                        propertyValue,
                        forceType
                    )
                }?.toMutableSet() ?: mutableSetOf()
            ).build()
        }.toTypedArray()
        PianoAnalytics.getInstance().sendEvents(*events)
    }

    private fun handleGetUser(): Map<String, Any?>? =
        PianoAnalytics.getInstance().userStorage.currentUser?.let {
            mapOf(
                "id" to it.id,
                "category" to it.category
            )
        }

    private fun handleSetUser(call: MethodCall) {
        PianoAnalytics.getInstance().userStorage.currentUser = User(
            call.arg("id"),
            category = call.argument("category"),
            shouldBeStored = call.argument<Boolean?>("enableStorage") ?: true
        )
    }

    private fun handleGetVisitorId(): String? =
        PianoAnalytics.getInstance().let {
            if (visitorIDType == VisitorIDType.CUSTOM)
                it.customVisitorId
            else
                it.visitorId
        }

    private fun handleSetVisitorId(call: MethodCall) {
        PianoAnalytics.getInstance().customVisitorId = call.arg("visitorId")
    }

    private fun privacyChangeStorageFeatures(call: MethodCall, include: Boolean = true) {
        val features = call.arg<List<String>>("features").map { feature ->
            PrivacyStorageFeature.entries.first { it.name == feature }
        }
        call.arg<List<String>>("modes").forEach {
            val mode = getPrivacyMode(it)
            (if (include) mode.allowedStorageFeatures else mode.forbiddenStorageFeatures) += features
        }
    }

    private fun privacyChangeProperties(call: MethodCall, include: Boolean = true) {
        val propertyNames = call.arg<List<String>>("propertyNames")
        val eventNames = call.argument<List<String>>("eventNames") ?: listOf(Event.ANY)

        call.arg<List<String>>("modes").forEach {
            val mode = getPrivacyMode(it)
            val propertyKeys = if (include) mode.allowedPropertyKeys else mode.forbiddenPropertyKeys

            eventNames.forEach { eventName ->
                propertyKeys
                    .getOrPut(eventName) { mutableSetOf() }
                    .addAll(propertyNames.map { name -> PropertyName(name) })
            }
        }
    }

    private fun privacyChangeEvents(call: MethodCall, include: Boolean = true) {
        val eventNames = call.arg<List<String>>("eventNames")

        call.arg<List<String>>("modes").forEach {
            val mode = getPrivacyMode(it)
            (if (include) mode.allowedEventNames else mode.forbiddenEventNames).addAll(eventNames)
        }
    }

    companion object {

        @JvmStatic
        private fun getPianoConsents() = try {
            PianoConsents.getInstance()
        } catch (_: IllegalStateException) {
            null
        }

        @JvmStatic
        private fun getVisitorIDType(name: String) = when (name) {
            "ADID" -> VisitorIDType.ADVERTISING_ID
            "CUSTOM" -> VisitorIDType.CUSTOM
            else -> VisitorIDType.UUID
        }

        @JvmStatic
        private fun getVisitorStorageMode(name: String?) = when (name) {
            "relative" -> VisitorStorageMode.RELATIVE
            else -> VisitorStorageMode.FIXED
        }

        @JvmStatic
        private fun getProperty(name: String, value: Any?, forceType: Property.Type? = null) =
            when (value) {
                is Boolean -> Property(PropertyName(name), value, forceType)
                is Int -> Property(PropertyName(name), value, forceType)
                is Long -> Property(PropertyName(name), value, forceType)
                is Double -> Property(PropertyName(name), value, forceType)
                is String -> Property(PropertyName(name), value, forceType)
                is Date -> Property(PropertyName(name), value)
                is List<*> -> {
                    when (value.first()) {
                        is Int -> Property(
                            PropertyName(name),
                            value.filterIsInstance<Int>().toTypedArray(),
                            forceType
                        )

                        is Double -> Property(
                            PropertyName(name),
                            value.filterIsInstance<Double>().toTypedArray(),
                            forceType
                        )

                        is String -> Property(
                            PropertyName(name),
                            value.filterIsInstance<String>().toTypedArray(),
                            forceType
                        )

                        else -> {
                            error("Invalid array value type of property \"$name\"")
                        }
                    }
                }

                else -> error("Invalid type of property \"$name\"")
            }

        @JvmStatic
        private fun getPrivacyMode(name: String) = when (name) {
            "opt-in" -> PrivacyMode.OPTIN
            "opt-out" -> PrivacyMode.OPTOUT
            "exempt" -> PrivacyMode.EXEMPT
            "custom" -> PrivacyMode.CUSTOM
            "no-consent" -> PrivacyMode.NO_CONSENT
            "no-storage" -> PrivacyMode.NO_STORAGE
            else -> error("Invalid privacy mode \"$name\"")
        }

    }
}

class MutableReportUrlProvider(
    collectDomain: String,
    site: Int,
    path: String = Configuration.DEFAULT_PATH
) : ReportUrlProvider {

    @Volatile
    private var _collectDomain = collectDomain
    @Volatile
    private var _site = site
    @Volatile
    private var _path = path

    override val collectDomain: String
        get() = _collectDomain
    override val site: Int
        get() = _site
    override val path: String
        get() = _path

    fun update(
        collectDomain: String = _collectDomain,
        site: Int = _site,
        path: String = _path
    ) {
        _collectDomain = collectDomain
        _site = site
        _path = path
    }
}